libr/anal/block.c

```diff
@@ -543,3 +558,246 @@ R_API R_NULLABLE RList/*<RAnalBlock *>*/ *r_anal_block_shortest_path(RAnalBlock
+		return block;
+	}
+	r_anal_block_ref (block);
+
+	// Cache all recursive successors of block here.
+	// These are the candidates that we might have to remove from functions later.
+	HtUP *succs = ht_up_new (NULL, noreturn_successor_free, NULL); // maps block addr (ut64) => NoreturnSuccessor *
+	if (!succs) {
+		return block;
+	}
+	r_anal_block_recurse (block, noreturn_successors_cb, succs);
+
+	// Chop the block. Resize and remove all destination addrs
+	r_anal_block_set_size (block, addr - block->addr);
+	block->jump = UT64_MAX;
+	block->fail = UT64_MAX;
+	r_anal_switch_op_free (block->switch_op);
+	block->switch_op = NULL;
+
+	// Now, for each fcn, check which of our successors are still reachable in the function remove and the ones that are not.
+	RListIter *it;
+	RAnalFunction *fcn;
+	r_list_foreach (block->fcns, it, fcn) {
+		RAnalBlock *entry = r_anal_get_block_at (block->anal, fcn->addr);
+		if (entry && r_list_contains (entry->fcns, fcn)) {
+			r_anal_block_recurse (entry, noreturn_successors_reachable_cb, succs);
+		}
+		ht_up_foreach (succs, noreturn_remove_unreachable_cb, fcn);
+	}
+
+	// This last step isn't really critical, but nice to have.
+	// Prepare to merge blocks with their predecessors if possible
+	RList merge_blocks;
+	r_list_init (&merge_blocks);
+	merge_blocks.free = (RListFree)r_anal_block_unref;
+	ht_up_foreach (succs, noreturn_get_blocks_cb, &merge_blocks);
+
+	// Free/unref BEFORE doing the merge!
+	// Some of the blocks might not be valid anymore later!
+	r_anal_block_unref (block);
+	ht_up_free (succs);
+
+	ut64 block_addr = block->addr; // save the addr to identify the block. the automerge might free it so we must not use the pointer!
+
+	// Do the actual merge
+	r_anal_block_automerge (&merge_blocks);
+
+	// No try to recover the pointer to the block if it still exists
+	RAnalBlock *ret = NULL;
+	for (it = merge_blocks.head; it && (block = it->data, 1); it = it->n) {
+		if (block->addr == block_addr) {
+			// block is still there
+			ret = block;
+			break;
+		}
+	}
+
+	r_list_purge (&merge_blocks);
+	return ret;
+}
+
+typedef struct {
+	HtUP *predecessors; // maps a block to its predecessor if it has exactly one, or NULL if there are multiple or the predecessor has multiple successors
+	HtUP *visited_blocks; // during predecessor search, mark blocks whose successors we already checked. Value is void *-casted count of successors
+	HtUP *blocks; // adresses of the blocks we might want to merge with their predecessors => RAnalBlock *
+
+	RAnalBlock *cur_pred;
+	size_t cur_succ_count;
+} AutomergeCtx;
+
+static bool automerge_predecessor_successor_cb(ut64 addr, void *user) {
+	AutomergeCtx *ctx = user;
+	ctx->cur_succ_count++;
+	RAnalBlock *block = ht_up_find (ctx->blocks, addr, NULL);
+	if (!block) {
+		// we shouldn't merge this one so GL_DONT_CARE
+		return true;
+	}
+	bool found;
+	RAnalBlock *pred = ht_up_find (ctx->predecessors, (ut64)block, &found);
+	if (!found) {
```

Comment:
be positive,use if (found) { first.

also. if ht_up_find fails always sets found? otherwse may be UB

Reply:
Yes, it always sets found. That is the sole purpose of the found parameter. To distinguish between `NULL` that is assigned to the given key and `NULL` when the key does not exist.